### PR TITLE
BUG: list-like to_replace on Categorical.replace is ignored or crash

### LINKED ISSUE #31720: `Series.replace` fails on categoricals with list
Looks like this was introduced in #27026: not sure of the details of what was incorrect in the old version but I do know that this at least didn't error:
```python
cat = pd.CategoricalIndex(['a', 'b']).to_series()
cat.replace('a', 'A')  # works
cat.replace(['a'], 'A')
~/model/.venv/lib/python3.7/site-packages/pandas/core/arrays/categorical.py in replace(self, to_replace, value, inplace)
   2440         inplace = validate_bool_kwarg(inplace, "inplace")
   2441         cat = self if inplace else self.copy()
-> 2442         if to_replace in cat.categories:
   2443             if isna(value):
   2444                 cat.remove_categories(to_replace, inplace=True)

~/model/.venv/lib/python3.7/site-packages/pandas/core/indexes/base.py in __contains__(self, key)
   3898     @Appender(_index_shared_docs["contains"] % _index_doc_kwargs)
   3899     def __contains__(self, key) -> bool:
-> 3900         hash(key)
   3901         try:
   3902             return key in self._engine

TypeError: unhashable type: 'list'
```
Seems like most of the methods used in this implementation already handle list-like inputs so it should be a pretty easy fix..?

#### Output of ``pd.show_versions()``

<details>
pandas           : 1.0.0
</details>

### PR DESCRIPTION
**Pull request by JustinZhengBC**
- [X] closes #31720
- [X] tests added / passed
- [x] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

Covers the case where `to_replace` is a list-like and `value` is a string. Other cases, like "`to_replace` is dict and `value` is None", or "`to_replace` and `value` are both lists" are handled earlier in generic.py

### CODE REVIEW

File: pandas/tests/arrays/categorical/test_replace.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import pandas as pd
+import pandas.testing as tm
```
Comment (MarcoGorelli) [inline pandas/tests/arrays/categorical/test_replace.py]: Thanks @JustinZhengBC 

This should be 
```
import pandas._testing as tm
```

**Comment by dsaxton** [comment]
How does this behave when `to_replace` is already a dict? On master the replacement seems to work but the data type changes:

```python
In [5]: s                                                                                                                                                             
Out[5]: 
0    0
1    1
2    2
dtype: category
Categories (3, int64): [0, 1, 2]

In [6]: s.replace({0: 1})                                                                                                                                             
Out[6]: 
0    1
1    1
2    2
dtype: int64
```

**Comment by JustinZhengBC** [comment]
@dsaxton this is because in `NDFrame` (generics.py) it chooses to not dispatch the dict case to the `self._data.replace`, but `self._data.replace_list` which takes a more complex path to iterate over its blocks and would need more drastic refactoring (it pre-computes masks to avoid repeated replacement if the same value appears more than once in `to_replace` or `value`). I think that is outside the scope of the current ticket

**Comment by MarcoGorelli** [comment]
Thanks @JustinZhengBC !

File: pandas/tests/arrays/categorical/test_replace.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+
+@pytest.mark.parametrize(
+    "to_replace,value,expected,check_types",
+    [
+        # one-to-one
+        (1, 2, [2, 2, 3], True),
+        (1, 4, [4, 2, 3], True),
+        (4, 1, [1, 2, 3], True),
+        (5, 6, [1, 2, 3], True),
```
Comment (MarcoGorelli) [inline pandas/tests/arrays/categorical/test_replace.py]: Does this need testing with different dtypes? e.g. with `to_replace=1`, `value="4"`

File: pandas/tests/arrays/categorical/test_replace.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        ([1, 2], 3, [3, 3, 3], True),
+        ([1, 2], 4, [4, 4, 3], True),
+        ((1, 2, 4), 5, [5, 5, 3], True),
+        ((5, 6), 2, [1, 2, 3], True),
+        # many-to-many, handled outside of Categorical and results in separate dtype
+        ([1], [2], [2, 2, 3], False),
+        ([1, 4], [5, 2], [5, 2, 3], False),
+    ],
+)
+def test_replace(to_replace, value, expected, check_types):
```
Comment (MarcoGorelli) [inline pandas/tests/arrays/categorical/test_replace.py]: Add the issue number here

File: pandas/_testing.py [hunk 4]
```diff
@@ -1104,6 +1105,8 @@ def assert_series_equal(
         Compare datetime-like which is comparable ignoring dtype.
     check_categorical : bool, default True
         Whether to compare internal Categorical exactly.
+    check_category_order : bool, default True
+        Whether to compare category order internal Categoricals
```
Comment (MarcoGorelli) [inline pandas/_testing.py]: Is there a missing "of" here, so
```
Whether to compare category order of internal Categoricals
```
?

File: pandas/core/arrays/categorical.py [hunk 5]
```diff
@@ -2412,6 +2412,9 @@ def isin(self, values):
         code_values = code_values[null_mask | (code_values >= 0)]
         return algorithms.isin(self.codes, code_values)
 
+    def replace_list(self, to_replace, value, inplace: bool = False):
+        return self.replace(to_replace, value, inplace)
+
```
Comment (MarcoGorelli) [inline pandas/core/arrays/categorical.py]: Where is this used?

File: pandas/core/arrays/categorical.py [hunk 6]
```diff
@@ -2441,19 +2444,29 @@ def replace(self, to_replace, value, inplace: bool = False):
         """
         inplace = validate_bool_kwarg(inplace, "inplace")
         cat = self if inplace else self.copy()
-        if to_replace in cat.categories:
-            if isna(value):
-                cat.remove_categories(to_replace, inplace=True)
+        if is_list_like(to_replace):
+            if is_list_like(value):
+                if len(to_replace) != len(value):
+                    raise ValueError("to_replace and value must be same length!")
```
Comment (MarcoGorelli) [inline pandas/core/arrays/categorical.py]: Could this be a bit more precise? As in, include something about the lengths of `to_replace` and `value` which were received

File: pandas/tests/arrays/categorical/test_replace.py (hunk 2 above)
Comment (JustinZhengBC) [reply pandas/tests/arrays/categorical/test_replace.py]: It's doable, with limited dtype-checking as sorting the category index causes a crash on comparison on mixed dtypes

File: pandas/tests/arrays/categorical/test_replace.py (hunk 3 above)
Comment (JustinZhengBC) [reply pandas/tests/arrays/categorical/test_replace.py]: fixed

File: pandas/_testing.py (hunk 4 above)
Comment (JustinZhengBC) [reply pandas/_testing.py]: fixed

File: pandas/core/arrays/categorical.py (hunk 5 above)
Comment (JustinZhengBC) [reply pandas/core/arrays/categorical.py]: sorry, this and the change below it were experiments in addressing the dict case and list-list case for categoricals, which turned out to be more complex than expected and are not intended to be in this pr

File: pandas/core/arrays/categorical.py (hunk 6 above)
Comment (JustinZhengBC) [reply pandas/core/arrays/categorical.py]: sorry, this and the change above it were experiments in addressing the dict case and list-list case for categoricals, which turned out to be more complex than expected and are not intended to be in this pr

File: pandas/tests/arrays/categorical/test_replace.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        ((5, 6), 2, [1, 2, 3], True, True),
+        # many-to-many, handled outside of Categorical and results in separate dtype
+        ([1], [2], [2, 2, 3], False, False),
+        ([1, 4], [5, 2], [5, 2, 3], False, False),
+        # check_categorical sorts categories, which crashes on mixed dtypes
+        (3, "4", [1, 2, "4"], True, False),
+        ([1, 2, "3"], "5", ["5", "5", 3], True, False),
+    ],
+)
+# GH 31720
```
Comment (MarcoGorelli) [inline pandas/tests/arrays/categorical/test_replace.py]: I think the issue number usually goes inside the test:
```
def test_me(*args):
    # GH 1234
```

Probably not worth changing on its own, but if you get any other comments requesting changes, perhaps this is worth doing

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.1.0.rst [hunk 8]
```diff
@@ -99,7 +99,7 @@ Bug fixes
 Categorical
 ^^^^^^^^^^^
 
--
+- Bug in :class:`Categorical` that would ignore or crash when calling :meth:`Series.replace` with a list-like ``to_replace`` (:issue:`31720`)
```
Comment (MarcoGorelli) [inline doc/source/whatsnew/v1.1.0.rst]: The issue is tagged with v1.0.2 milestone, I think the should go in that whatsnew file

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/categorical.py [hunk 9]
```diff
@@ -2441,19 +2441,24 @@ def replace(self, to_replace, value, inplace: bool = False):
...
-            if isna(value):
-                cat.remove_categories(to_replace, inplace=True)
-            else:
-                categories = cat.categories.tolist()
-                index = categories.index(to_replace)
-                if value in cat.categories:
-                    value_index = categories.index(value)
-                    cat._codes[cat._codes == index] = value_index
-                    cat.remove_categories(to_replace, inplace=True)
+        if is_list_like(to_replace):
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: can you add some comments here on the procure for what you are doing.

File: pandas/_testing.py [hunk 10]
```diff
@@ -1104,6 +1105,8 @@ def assert_series_equal(
         Compare datetime-like which is comparable ignoring dtype.
     check_categorical : bool, default True
         Whether to compare internal Categorical exactly.
+    check_category_order : bool, default True
```
Comment (jreback) [inline pandas/_testing.py]: can we simply enforce ordering when check_categorical is True? alteratively, could add 'exact' as a possiblity in check_categorical (just see what is easier here, not averse to adding another arg)

**Review by jreback — CHANGES_REQUESTED** [review]
looks pretty good. move the note to 1.0.2

File: pandas/_testing.py (hunk 10 above)
Comment (JustinZhengBC) [reply pandas/_testing.py]: I think adding the extra arg is easiest. `pytest.mark.parametrize` doesn't work with optional arguments so there's no way to use optional arguments to cut down on args needed. Adding a string possibility to a boolean argument isn't very intuitive (and changing `check_categorical` to a string argument would result in a lot of changes in other files)

File: pandas/core/arrays/categorical.py (hunk 9 above)
Comment (JustinZhengBC) [reply pandas/core/arrays/categorical.py]: fixed, added comments around `replace_dict` generation

File: doc/source/whatsnew/v1.1.0.rst (hunk 8 above)
Comment (JustinZhengBC) [reply doc/source/whatsnew/v1.1.0.rst]: fixed

File: pandas/tests/arrays/categorical/test_replace.py (hunk 7 above)
Comment (JustinZhengBC) [reply pandas/tests/arrays/categorical/test_replace.py]: fixed

**Comment by JustinZhengBC** [comment]
@jreback I've moved the whatsnew and addressed the other feedback

File: pandas/core/arrays/categorical.py [hunk 11]
```diff
@@ -2441,19 +2441,31 @@ def replace(self, to_replace, value, inplace: bool = False):
...
+            replace_dict = {replace_value: value for replace_value in to_replace}
+        else:
+            # if both to_replace and value are scalar
+            replace_dict = {to_replace: value}
+        # other cases, like if both to_replace and value are list-like or if
+        # to_replace is a dict, are handled separately in NDFrame
+
+        for replace_value, new_value in replace_dict.items():
+            if replace_value in cat.categories:
+                if isna(new_value):
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: move the isna check first as it an never be true by definition (no nulls in categories); also simplifies the logic slightly

```

if isna(….):
    ….
   continue

….rest of logic
```

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
also pls merge master

File: pandas/core/arrays/categorical.py (hunk 11 above)
Comment (JustinZhengBC) [reply pandas/core/arrays/categorical.py]: @jreback This breaks a lot of other tests. I think isna should be allowed. The [documentation](https://pandas.pydata.org/pandas-docs/stable/user_guide/categorical.html) says "All values of categorical data are either in categories or np.nan"

PS I merged master

Comment (jreback) [reply pandas/core/arrays/categorical.py]: you are missing my point. you simply need to reorganize, its the same condition w/o the dedent

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/categorical.py [hunk 12]
```diff
@@ -2440,18 +2440,30 @@ def replace(self, to_replace, value, inplace: bool = False):
...
-            else:
+
+        # build a dict of (to replace -> value) pairs
+        if is_list_like(to_replace):
+            # if to_replace is list-like and value is scalar
+            replace_dict = {replace_value: value for replace_value in to_replace}
+        else:
+            # if both to_replace and value are scalar
+            replace_dict = {to_replace: value}
+        # other cases, like if both to_replace and value are list-like or if
```
Comment (jreback) [inline pandas/core/arrays/categorical.py]: can you put a blank line here (before other cases), and remove the one before the for loop. Its just confusing where the comment applies now

File: pandas/_testing.py (hunk 10 above)
Comment (jreback) [inline pandas/_testing.py]: can you add a versionadded tag here (1.0.2)

**Review by jreback — CHANGES_REQUESTED** [review]
very minor comments, ping on green.

**Comment by JustinZhengBC** [comment]
@jreback green

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @JustinZhengBC very nice!

**Pull request merged**